Item 7095: Add detailed audit logging for samples as separate audit type event

labkey-susanh opened the pull request:
**Rationale**
For the sample timeline feature within LabKey Sample Manager, we require detailed audit logs for samples that can be more easily reference by sample Id.  The current placement of these detailed logs in the QueryUpdateEvents makes it difficult to find relevant logs and also doesn't capture all the desired metadata.  We also want to have this detailed audit logging turned on for Sample Manager without requiring that it be turned on for all Sample Types.

**Related PRs**
- https://github.com/LabKey/labkey-api-js/pull/52
- https://github.com/LabKey/labkey-ui-components/pull/220
- https://github.com/LabKey/sampleManagement/pull/235

**Changes**
- Refactor handling and display of detailed audit records for better reuse
- Allow some APIs to override auditBehaviorType via a parameter 
- Add new Audit Provider, domain and event type for storing detailed logs for Samples, replacing their inclusion in the QueryUpdateEvents audit log.

labkey-susanh committed:
Add SampleTimelineAuditProvider

labkey-susanh committed:
Merge remote-tracking branch 'origin/develop' into fb_sampleTimelineAuditEvents

labkey-susanh committed:
spelling

labkey-susanh committed:
Refactor detailed audit logging to separate logging for SampleTimeline from general QueryUpdateEvent logging

labkey-susanh committed:
Some field renaming and refactor of addAuditEvent for use in all types of events, distinguishing sample set audit events from query update events

labkey-susanh committed:
Move action for showing a detailed audit log page to the AuditController for better reuse

labkey-susanh committed:
Helps when you add the detail fields to the domain...

labkey-susanh committed:
Customize the audit log comments

labkey-susanh committed:
Spelling

labkey-susanh committed:
Set title column so it isn't the comment field

labkey-susanh committed:
Remove commented code

labkey-susanh committed:
Merge remote-tracking branch 'origin/develop' into fb_sampleTimelineAuditEvents

labkey-susanh committed:
Refactoring to allow flagging lineage changes for sample events

labkey-susanh committed:
Add metadata field for sample timeline event

labkey-susanh committed:
Merge remote-tracking branch 'origin/develop' into fb_sampleTimelineAuditEvents

labkey-susanh committed:
Add sampleSet Id as key in domain but leave LSID until we get the DBSequence in place

labkey-susanh committed:
Add detailed delete audit event if appropriate

labkey-susanh committed:
Merge remote-tracking branch 'origin/develop' into fb_sampleTimelineAuditEvents

XingY commented on api/src/org/labkey/api/audit/DetailedAuditHandler.java:
```diff
@@ -0,0 +1,151 @@
+import java.util.Objects;
+import java.util.Set;
+
+public abstract class DetailedAuditHandler
```
I find the class name a bit misleading... It seems to handle all audit types, not just detailed

labkey-susanh replied:
True. The addAuditEvent specifically handles adding the oldRecord and newRecord fields, which are available only in a DetailedAuditTypeEvent, and that's the main part I was trying to avoid duplicating.  I can rename to AuditHandler.  It seems maybe too general, but we'll see.

XingY commented on api/src/org/labkey/api/audit/DetailedAuditHandler.java:
```diff
@@ -0,0 +1,151 @@
+        if (table.supportsAuditTracking())
+        {
+            AuditConfigurable auditConfigurable = (AuditConfigurable)table;
+            auditType = auditType == null ? auditConfigurable.getAuditBehavior() : auditType;
```
With this approach, will sample manager still able to disable detailed logging via xml config?

labkey-susanh replied:
No, but that is intentional.  All actions that come from the sample manager application should use detailed logging, otherwise we have no timeline feature to speak of.  This also follows the general practice for command line tools and precedences, where the options set on the command line have highest precedence (I consider the URL parameters to be analogous to the command-line options.)   

For performance testing purposes, we can possibly have an admin flag that can be set to turn off the detailed logging (and right now we can use the experimental flag created in your branch), but this isn't something a LSKM user will have access to.

labkey-jeckels replied:
I don't think we should allow the URL to override what an admin set up for logging. A malicious user shouldn't be able to opt out of audit logging.

labkey-susanh replied:
Ah, yes.  Malicious users.  OK, I'll make this change.  I predict there will be some difficulties with this timeline feature because of the different ways it can be affected.

XingY commented on experiment/src/org/labkey/experiment/api/SampleSetServiceImpl.java:
```diff
@@ -870,4 +884,101 @@ public boolean parentAliasHasCorrectFormat(String parentAlias)
+            String sampleSetLsid = null;
+            if (row.containsKey(CPAS_TYPE))
+                sampleSetLsid =  String.valueOf(row.get(CPAS_TYPE));
+            if (sampleSetLsid == null && row.containsKey("sampleset"))
```
maybe a comment on when sampleset is the lsid

labkey-susanh replied:
Good thought.  Now if I can just remember when that happens...

XingY commented on experiment/src/org/labkey/experiment/api/SampleSetServiceImpl.java:
```diff
@@ -870,4 +884,101 @@ public boolean parentAliasHasCorrectFormat(String parentAlias)
+            case DELETE:
+            case TRUNCATE:
+                return "Sample was deleted.";
+            case MERGE:
```
Since the old/new data map objects are available, we should be able to determine if a single row is register or update.

XingY replied:
OK nvm, I see that merge doesn't have old data map.

XingY reviewed (commented):
My concern is that a SM user will not be able to turn detailed logging off via XML. Maybe auditConfigurable should be preferred over context param? Looks good otherwise.

labkey-susanh committed:
Merge remote-tracking branch 'origin/develop' into fb_sampleTimelineAuditEvents

labkey-susanh committed:
Add auditBehavior parameter (mirroring change in labkey-api-js) to allow override of audit behavior

labkey-susanh committed:
DetailedAuditHandler -> AuditHandler
Make XML audit behavior override command line behavior
Use MERGE action when InsertOption is set to MERGE

labkey-susanh committed:
Add comment

labkey-susanh committed:
Set sample ID on audit event

XingY commented on api/src/org/labkey/api/audit/AuditHandler.java:
```diff
@@ -36,7 +36,8 @@ public void addAuditEvent(User user, Container c, TableInfo table, @Nullable Aud
         {
             AuditConfigurable auditConfigurable = (AuditConfigurable)table;
-            auditType = auditType == null ? auditConfigurable.getAuditBehavior() : auditType;
+            if (auditType == null || auditConfigurable.getXmlAuditBehaviorType() != null)
```
What if both auditType and auditConfigurable.getXmlAuditBehaviorType are null, but auditConfigurable.getAuditBehavior is not null? I think getAuditBehavior is never null, since it defaults to NONE. But it can also be not NONE. DataClass has an override that would set default to DETAILED, although getXmlAuditBehaviorType is null.

labkey-susanh replied:
If both are null, we'll use the result of getAuditBehavior, whatever that is (and I don't think it can be null).  Isn't that what we want?

XingY replied:
You are right, I misunderstood the if condition.

labkey-susanh committed:
Move SampleTimelineAuditEvent to API for better access

labkey-susanh committed:
Merge remote-tracking branch 'origin/develop' into fb_sampleTimelineAuditEvents
